src/contractor/contractor.cpp

```diff
@@ -172,37 +185,74 @@ std::size_t Contractor::LoadEdgeExpandedGraph(
-            while (csv_in.read_row(from_node_id, to_node_id, speed))
+            for (auto segment_speed_filename : segment_speed_filenames)
             {
-                segment_speed_lookup[std::make_pair(OSMNodeID(from_node_id),
-                                                    OSMNodeID(to_node_id))] =
-                    std::make_pair(speed, file_id);
+                util::SimpleLogger().Write()
+                    << "Segment speed data supplied, will update edge weights from "
+                    << segment_speed_filename;
+                io::CSVReader<3> csv_in(segment_speed_filename);
+                csv_in.set_header("from_node", "to_node", "speed");
+                uint64_t from_node_id{};
+                uint64_t to_node_id{};
+                unsigned speed{};
+                while (csv_in.read_row(from_node_id, to_node_id, speed))
+                {
+                    segment_speed_lookup[std::make_pair(OSMNodeID(from_node_id),
+                                                        OSMNodeID(to_node_id))] =
+                        std::make_pair(speed, segment_file_id);
+                }
+                ++segment_file_id;
+
+                // Check for overflow
+                if (segment_file_id == 0)
+                {
+                    throw util::exception(
+                        "Sorry, there's a limit of 254 segment speed files; you supplied too many");
+                }
             }
-            ++file_id;
+        }
 
-            // Check for overflow
-            if (file_id == 0)
+        if (update_turn_penalties)
+        {
+            for (auto turn_penalty_filename : turn_penalty_filenames)
             {
-                throw util::exception(
-                    "Sorry, there's a limit of 254 segment speed files, you supplied too many");
+                util::SimpleLogger().Write()
+                    << "Turn penalty data supplied, will update turn penalties from "
+                    << turn_penalty_filename;
+                io::CSVReader<4> csv_in(turn_penalty_filename);
+                csv_in.set_header("from_node", "via_node", "to_node", "penalty");
+                uint64_t from_node_id{};
+                uint64_t via_node_id{};
+                uint64_t to_node_id{};
+                double penalty{};
+                while (csv_in.read_row(from_node_id, via_node_id, to_node_id, penalty))
+                {
+                    turn_penalty_lookup[std::make_tuple(
+                        OSMNodeID(from_node_id), OSMNodeID(via_node_id), OSMNodeID(to_node_id))] =
+                        std::make_pair(penalty, turn_file_id);
+                }
+                ++turn_file_id;
+
+                // Check for overflow
+                if (turn_file_id == 0)
+                {
+                    throw util::exception(
+                        "Sorry, there's a limit of 254 turn penalty files; you supplied too many");
+                }
             }
         }
 
```

Comment:
looks like the `if( update...)` loop should end here? it includes the full a lot of stuff that shouldn't be in here, as well?